examples/cpp/CMakeLists.txt

```diff
@@ -6,8 +6,8 @@ include_directories(${PROJECT_SOURCE_DIR}/src/cc)
 include_directories(${PROJECT_SOURCE_DIR}/src/cc/api)
 include_directories(${PROJECT_SOURCE_DIR}/src/cc/libbpf/include/uapi)
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fPIC")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC -shared")
```

Comment:
Pleas also check if `cmake -DENABLE_LLVM_SHARED=1 ..` fix the problem.

Reply:
-D did not work:

```
$ cmake -DENABLE_LLVM_SHARED=1 ..
...
$ make -j4
...
[ 81%] Linking CXX shared library libbcc.so
/usr/bin/ld: cannot find -lLLVM
collect2: error: ld returned 1 exit status
make[2]: *** [src/cc/CMakeFiles/bcc-shared.dir/build.make:344: src/cc/libbcc.so.0.25.0] Error 1
make[1]: *** [CMakeFiles/Makefile2:852: src/cc/CMakeFiles/bcc-shared.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 81%] Building CXX object src/cc/CMakeFiles/bcc-static.dir/usdt/usdt.cc.o
[ 81%] Building CXX object src/cc/CMakeFiles/bcc-static.dir/usdt/usdt_args.cc.o
[ 82%] Building CXX object src/cc/CMakeFiles/bcc-static.dir/bcc_syms.cc.o
[ 82%] Building C object src/cc/CMakeFiles/bcc-static.dir/bcc_elf.c.o
[ 82%] Building C object src/cc/CMakeFiles/bcc-static.dir/bcc_perf_map.c.o
[ 83%] Building C object src/cc/CMakeFiles/bcc-static.dir/bcc_proc.c.o
[ 83%] Linking CXX static library libbcc.a
[ 83%] Built target bcc-static
make: *** [Makefile:146: all] Error 2
```

Reply:
I don't get this problem with command line

```
cmake \
	-DCMAKE_INSTALL_PREFIX=/usr \
	-DENABLE_LLVM_SHARED=1 \
	-DCMAKE_BUILD_TYPE=Debug \
	..

make -j$(nproc)
```

Is this problem belongs to `error adding symbols: DSO missing from command line`?

Reply:
> I don't get this problem with command line
> 
> ```
> cmake \
> 	-DCMAKE_INSTALL_PREFIX=/usr \
> 	-DENABLE_LLVM_SHARED=1 \
> 	-DCMAKE_BUILD_TYPE=Debug \
> 	..
> 
> make -j$(nproc)
> ```
> 
> Is this problem belongs to `error adding symbols: DSO missing from command line`?

For me it ran into a secondary issue, `ld` not finding `libLLVM.so` (my 14.0.6 install does not have such lib, but instead a bunch of LLVM libraries).

Verbose output (please don't mind LLVM 13, I'm doing this on multiple machines):
```
cd /home/rnsanchez/bcc-copy/build/src/cc && /usr/bin/cmake -E cmake_link_script CMakeFiles/bcc-shared.dir/link.txt --verbose=1
/usr/bin/c++ -fPIC  -Wall  -fno-rtti -fPIC -DBCC_PROG_TAG_DIR='"/var/tmp/bcc"' -Wno-unused-result -DLLVM_MAJOR_VERSION=13 -O3 -DNDEBUG   -Wl,--exclude-libs=libclang-cpp.so  -Wl,--exclude-libs=libLLVM.a -shared -Wl,-soname,libbcc.so.0 -o lib
bcc.so.0.25.0 CMakeFiles/bcc-shared.dir/link_all.cc.o CMakeFiles/bcc-shared.dir/bcc_common.cc.o CMakeFiles/bcc-shared.dir/bpf_module.cc.o CMakeFiles/bcc-shared.dir/bcc_btf.cc.o CMakeFiles/bcc-shared.dir/exported_files.cc.o CMakeFiles/bcc-sh
ared.dir/bcc_debug.cc.o CMakeFiles/bcc-shared.dir/bpf_module_rw_engine.cc.o CMakeFiles/bcc-shared.dir/table_storage.cc.o CMakeFiles/bcc-shared.dir/shared_table.cc.o CMakeFiles/bcc-shared.dir/bpffs_table.cc.o CMakeFiles/bcc-shared.dir/json_m
ap_decl_visitor.cc.o CMakeFiles/bcc-shared.dir/bcc_syms.cc.o CMakeFiles/bcc-shared.dir/bcc_elf.c.o CMakeFiles/bcc-shared.dir/bcc_perf_map.c.o CMakeFiles/bcc-shared.dir/bcc_proc.c.o CMakeFiles/bcc-shared.dir/common.cc.o  frontends/clang/libc
lang_frontend.a -Wl,--whole-archive -lclang-cpp -lLLVM -Wl,--no-whole-archive -lelf -ldebuginfod libbcc_bpf.a -Wl,--whole-archive api/libapi-static.a -Wl,--no-whole-archive usdt/libusdt-static.a -lelf -lz 
/usr/bin/ld: cannot find -lLLVM
collect2: error: ld returned 1 exit status
make[2]: *** [src/cc/CMakeFiles/bcc-shared.dir/build.make:344: src/cc/libbcc.so.0.25.0] Error 1
make[2]: Leaving directory '/home/rnsanchez/bcc-copy/build'
make[1]: *** [CMakeFiles/Makefile2:852: src/cc/CMakeFiles/bcc-shared.dir/all] Error 2
make[1]: Leaving directory '/home/rnsanchez/bcc-copy/build'
make: *** [Makefile:146: all] Error 2
```

Reply:
I can't get same problem in my another computer anyway, My compile detail:

```
build/tests/cc/CMakeFiles/test_static.dir/link.txt:1:/usr/bin/c++  -Wall  -Wno-unused-result -fPIC -g -rdynamic CMakeFiles/test_static.dir/test_static.c.o -o test_static  ../../src/cc/libbcc.a ../../src/cc/frontends/clang/libclang_frontend.a -Wl,--whole-archive -lclang-cpp /usr/lib64/libLLVM-14.so -Wl,--no-whole-archive -lelf -ldebuginfod ../../src/cc/libbcc_bpf.a ../../src/cc/api/libapi-static.a ../../src/cc/usdt/libusdt-static.a ../../src/cc/libbcc-loader-static.a -lelf -lz
```
link to `libLLVM-14.so `:
```
[rongtao@RT-PC bcc]$ ldconfig -p | grep libLLVM
	libLLVM-14.so (libc6,x86-64) => /lib64/libLLVM-14.so
```
Do you try `rm -rf build`:

```
$ cd build
$ rm -rf *
$ cmake -DENABLE_LLVM_SHARED=1 ..
```

Reply:
> link to `libLLVM-14.so `:
> 
> ```
> [rongtao@RT-PC bcc]$ ldconfig -p | grep libLLVM
> 	libLLVM-14.so (libc6,x86-64) => /lib64/libLLVM-14.so
> ```

Here's wildly different; I don't see any particular reason though, upon inspecting Slackware's build script for LLVM (but I should add I only rarely rebuild LLVM):
```
$ /sbin/ldconfig -p | grep libLLVM
	libLLVMipo.so.14 (libc6,x86-64) => /usr/lib64/libLLVMipo.so.14
	libLLVMipo.so (libc6,x86-64) => /usr/lib64/libLLVMipo.so
	libLLVMX86TargetMCA.so.14 (libc6,x86-64) => /usr/lib64/libLLVMX86TargetMCA.so.14
	libLLVMX86TargetMCA.so (libc6,x86-64) => /usr/lib64/libLLVMX86TargetMCA.so
	libLLVMX86Info.so.14 (libc6,x86-64) => /usr/lib64/libLLVMX86Info.so.14
	libLLVMX86Info.so (libc6,x86-64) => /usr/lib64/libLLVMX86Info.so
...
	libLLVMAMDGPUCodeGen.so (libc6,x86-64) => /usr/lib64/libLLVMAMDGPUCodeGen.so
	libLLVMAMDGPUAsmParser.so.14 (libc6,x86-64) => /usr/lib64/libLLVMAMDGPUAsmParser.so.14
	libLLVMAMDGPUAsmParser.so (libc6,x86-64) => /usr/lib64/libLLVMAMDGPUAsmParser.so
```

> Do you try `rm -rf build`, compile completely from zero:

Yes, I always remove the `build` directory entirely.

Reply:
> Check `bcc/cmake/clang_libs.cmake` L2:
> 
> ```
>   1 if(ENABLE_LLVM_SHARED)                                                          
>   2 set(llvm_libs "LLVM")                                                                                             
>   3 else()
> ```
> 
> libLLVM.so's name determined by `${llvm_libs}`, does `llvm` compiled by your self?

I've recompiled in order to preserve symbols; other than that just as it is compiled for an official Slackware64 package:

https://mirrors.slackware.com/slackware/slackware64-current/source/d/llvm/

But here (actually in Slackware) it is not libLLVM, it is split into multiple components/libraries.  I don't know why, but from the build script I don't see anything peculiar being done to end up with such organization of multiple libraries.  Line 2 above should not hardcode to "LLVM" but instead a shell equivalent of `llvm-config --link-shared --libs` with a possible prefix of `llvm-config --ldflags`.

> I think there maybe some method to solve this problem:
>     * install llvm with 'dnf/yum', 'apt-get', ...
>     * For example: In my OS, `/lib64/libLLVMipo.a` is static library, yours `/usr/lib64/libLLVMipo.so.14` is shared library.
>     * In my OS has `/lib64/libLLVM.so -> libLLVM-14.so`,

That does not work in Slackware.  :-)

> That is to say, `/usr/bin/ld: cannot find -lLLVM` is normal on your operating system?

Yes, because it is not built as a monolithic library.  But I can compile your test and force linking to LLVM libraries (added `-v` for extra verbosity):

```
$ clang t.cpp $(llvm-config --ldflags) $(llvm-config --link-shared --libs) -v
clang version 14.0.6
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0
Selected GCC installation: /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "/usr/bin/clang-14" -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -mrelax-all --mrelax-relocations -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name t.cpp -mrelocation-model static -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debugger-tuning=gdb -v -fcoverage-compilation-dir=/tmp/clang -resource-dir /usr/bin/../lib64/clang/14.0.6 -cxx-isystem /usr/lib64/qt/include -internal-isystem /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../include/c++/12.2.0 -internal-isystem /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../include/c++/12.2.0/x86_64-slackware-linux -internal-isystem /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../include/c++/12.2.0/backward -internal-isystem /usr/bin/../lib64/clang/14.0.6/include -internal-isystem /usr/local/include -internal-isystem /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../x86_64-slackware-linux/include -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdeprecated-macro -fdebug-compilation-dir=/tmp/clang -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -fcolor-diagnostics -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/t-9de945.o -x c++ t.cpp
clang -cc1 version 14.0.6 based upon LLVM 14.0.6 default target x86_64-unknown-linux-gnu
ignoring nonexistent directory "/usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../x86_64-slackware-linux/include"
ignoring nonexistent directory "/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib64/qt/include
 /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../include/c++/12.2.0
 /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../include/c++/12.2.0/x86_64-slackware-linux
 /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../include/c++/12.2.0/backward
 /usr/bin/../lib64/clang/14.0.6/include
 /usr/local/include
 /usr/include
End of search list.
 "/usr/bin/ld" --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o a.out /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../lib64/crt1.o /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../lib64/crti.o /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/crtbegin.o -L/usr/lib64 -L/usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0 -L/usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/lib -L/usr/lib /tmp/t-9de945.o -lLLVMWindowsManifest -lLLVMXRay -lLLVMLibDriver -lLLVMDlltoolDriver -lLLVMCoverage -lLLVMLineEditor -lLLVMX86TargetMCA -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMX86Desc -lLLVMX86Info -lLLVMBPFDisassembler -lLLVMBPFAsmParser -lLLVMBPFCodeGen -lLLVMBPFDesc -lLLVMBPFInfo -lLLVMAMDGPUTargetMCA -lLLVMAMDGPUDisassembler -lLLVMAMDGPUAsmParser -lLLVMAMDGPUCodeGen -lLLVMAMDGPUDesc -lLLVMAMDGPUUtils -lLLVMAMDGPUInfo -lLLVMOrcJIT -lLLVMMCJIT -lLLVMJITLink -lLLVMInterpreter -lLLVMExecutionEngine -lLLVMRuntimeDyld -lLLVMOrcTargetProcess -lLLVMOrcShared -lLLVMDWP -lLLVMSymbolize -lLLVMDebugInfoPDB -lLLVMDebugInfoGSYM -lLLVMOption -lLLVMObjectYAML -lLLVMMCA -lLLVMMCDisassembler -lLLVMLTO -lLLVMCFGuard -lLLVMFrontendOpenACC -lLLVMExtensions -lPolly -lPollyISL -lLLVMPasses -lLLVMObjCARCOpts -lLLVMCoroutines -lLLVMipo -lLLVMInstrumentation -lLLVMVectorize -lLLVMLinker -lLLVMFrontendOpenMP -lLLVMDWARFLinker -lLLVMGlobalISel -lLLVMMIRParser -lLLVMAsmPrinter -lLLVMDebugInfoMSF -lLLVMSelectionDAG -lLLVMCodeGen -lLLVMIRReader -lLLVMAsmParser -lLLVMInterfaceStub -lLLVMFileCheck -lLLVMFuzzMutate -lLLVMTarget -lLLVMScalarOpts -lLLVMInstCombine -lLLVMAggressiveInstCombine -lLLVMTransformUtils -lLLVMBitWriter -lLLVMAnalysis -lLLVMProfileData -lLLVMDebugInfoDWARF -lLLVMObject -lLLVMTextAPI -lLLVMMCParser -lLLVMMC -lLLVMDebugInfoCodeView -lLLVMBitReader -lLLVMCore -lLLVMRemarks -lLLVMBitstreamReader -lLLVMBinaryFormat -lLLVMTableGen -lLLVMSupport -lLLVMDemangle -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/crtend.o /usr/bin/../lib64/gcc/x86_64-slackware-linux/12.2.0/../../../../lib64/crtn.o
```

Reply:
Yes, i think your problem not belongs to `error adding symbols: DSO missing from command line`, may be you should check `/usr/lib64/cmake/llvm/LLVMConfig.cmake` in your system and submit another PR to `bcc/cmake/clang_libs.cmake`.

Reply:
`bcc` need -lLLVM here by default, 
or you can create a link like `/lib64/libLLVM.so -> libLLVM-YOUR-SHARED-LIBS.so` manually, and test it again?

Reply:
> Yes, i think your problem not belongs to `error adding symbols: DSO missing from command line`, may be you should check `/usr/lib64/cmake/llvm/LLVMConfig.cmake` in your system and submit another PR to `bcc/cmake/clang_libs.cmake`.

I will consider it, thanks!